### DISCUSSION TITLE
Abstraction vs Core

### DISCUSSION THREAD

**Discussion by johndcollins** [OrchardCMS/OrchardCore#13595] [Q&A]
Hi, I'm looking at building my own module. 
I'm looking at existing modules for guidance and I'm a little confused about the Abstractions and Core projects. Sometimes they seem to be used differently. Some have models in abstractions and some have them in cores, what's the thought around these two types?

**Comment by ns8482e** [comment]
Use abstraction when you have abstract classes or interfaces or models used in interfaces/abstract classes that might have implementation span across modules.

Use core for default implementations, models, services belong to one module. Core can also be used by more than one modules

**Comment by johndcollins** [comment]
Thanks @ns8482e 
There seems to be very little documentation on creating modules from scratch, except for with the old version. Does most of that still apply to OrchardCore?

**Comment by ns8482e** [comment]
Yes there is no such documentation as guided tour like it was in for Orchard CMS. However here's a starting point on how to create a cms module https://docs.orchardcore.net/en/latest/docs/getting-started/templates/#create-a-new-cms-module

**Comment by johndcollins** [comment]
Ok, thanks, that's not a lot.
Thanks for your feedback.

**Comment by hishamco** [comment]
@johndcollins you can have a look to OC source or [OCC ](https://github.com/OrchardCoreContrib/OrchardCoreContrib) source there are some abstractions and core APIs that might make it clear for you to distinguish between both

FYI ASP.NET Core APIs are using the same principle

**Comment by ns8482e** [comment]
My understanding is Orchard core source is following clean architecture style pattern where code for functional module is defined generally (not always) between Core,Infrastructure, Module.

Core
You can define Entities, Parts, Domain Services, Interfaces, Domain events etc

Infrastructure 
You can define things like data access, DataMigration, Deployment, Commands, Logger etc- OC uses this in limited module.

UI/Module
You can define controller, filters, Views, Drivers, ViewModels, features and Startup, middleware etc. 


Abstractions
Orchard core being modular application and a modular framework it also have some features like display management and content management, localization - here related libraries split between core and abstractions.

Some libraries require multiple implementations targeting different scenarios hence, some related interfaces are defined in abstractions and have multiple core implementation.


You do not require to define abstractions in your module if you are only anticipating single implementation for your abstract/interfaces then you can include them in Core.  

I hope it clarifies some between core and abstractions and when to define them for your module.
